### PR TITLE
PR 16: Implementation of Issue #2.4 - Work Location Feature

### LINKED ISSUE #4: Issue: Implement Backend Core & Frontend Base Architecture
## Description
Implement the foundational backend architecture using FastAPI and set up the initial frontend project structure with React + Vite. This task covers iteration-1 requirements including basic authentication, user management, and meal participation logic with JSON storage.

### Backend
- [x] **Project Structure**: Set up FastAPI app with modular routers (`auth`, `users`, `meals`).
- [x] **Authentication**: JWT-based Login and Registration.
- [x] **Data Models**: Pydantic models for `User` (Employee/Team Lead/Admin) and `MealParticipation`.
- [x] **Storage**: JSON file-based persistence layer.
- [x] **Business Logic**:
    - [x] Employees can mark attendance for Lunch/Snacks.
    - [x] System enforces a 9:00 PM cutoff time for next-day updates.
    - [x] Admin/Team Leads can view headcounts and override entries.

### Frontend
- [x] **Initialization**: Create Vite + React project.
- [x] **Routing**: Set up `react-router-dom`.
- [x] **Security**: Implement Protected Routes (RBAC) and Auth Context.

## Acceptance Criteria
- Backend API is running on port 8000 and accessible via Swagger UI (`/docs`).
- Users can register and login, receiving a valid JWT.
- Frontend application starts without errors and handles basic navigation based on user roles.

## Linked PRs
- #3

### LINKED ISSUE #23: Issue #2.4: Work Location for a Date
## Summary
Employees can set their work location (Office / WFH) for a selected date via the Discord bot. Team Leads and Admin can correct missing work-location entries.

## Context
Work location tracking is needed for headcount reporting (Office vs WFH split). Employees set their own location; TLs/Admins can fill in missing entries. This data feeds into headcount summaries and WFH usage tracking.

## Expected Behavior
- `/work-location` slash command with optional `date` parameter (defaults to today).
- Bot replies with the employee's current work location for the date and interactive buttons: "Office" / "WFH".
- Employee clicks a button to set or change their location.
- Bot updates DynamoDB and confirms the change.
- Same cutoff rules apply as meal participation.
- TLs can set work location for employees in their team (via `/override-update` or a dedicated sub-command).
- Admins can set work location for any employee.

## Technical Notes
- Work location is stored alongside meal participation in DynamoDB (same record or related item).
- Default location: Office (unless a company-wide WFH period is active — see Issue #30).
- WFH days count toward the monthly soft limit (see Issue #34).

## Resources
- Design doc: User flows → [Employee self-update](https://github.com/NiloyGabrielGomes/Tasks-Repo_CEA-2026-Niloy/blob/5ae6ca0185bd09d7026efdc692b300628c6d899a/Task2_mhp_discord-bot/MHP_Discord-Bot.md?plain=1#L81-L84)

### PR DESCRIPTION
**Pull request by NiloyGabrielGomes**
## Related Issues
* Fixes #23

## Summary
Implements the `/work-location` slash command and button interaction handlers, allowing employees to view and set their daily work location (Office or WFH) via Discord. Includes WFH monthly soft cap enforcement and shared validation rules with meal participation.

## Dependencies
- #42 (Issue #22 — Meal Participation code)
- #43 (Issue #23 — Work Location docs)

## Changes
- Added `location_service.py` — `LocationService` class with get/set location, WFH monthly cap check (`_check_wfh_cap`), per-day fallback scan, and display helpers
- Added `location_handler.py` — `/work-location` command handler, `location_set:*` button handler, embed builder with Office (green) / WFH (yellow) color coding, and Office/WFH toggle buttons with active-state styling
- Modified `interaction.py` — Wired `work-location` command routing and `location_set:` button prefix routing
- Added `tests/test_location_service.py` — 30 tests covering service logic, handler responses, WFH cap enforcement, date validation, wrong-user rejection, and edge cases

## Context
Issue #4 is the second employee-facing feature in the MHP Discord Bot. It follows the same architectural pattern as Issue #3 (Meal Participation): slash command → validate date → query DynamoDB → return ephemeral embed with buttons → handle button click → update DynamoDB → return UPDATE_MESSAGE. The WFH monthly cap (default 5 days/month, configurable via `WFH_MONTHLY_CAP`) is a new validation layer specific to work location.

## How to Test (if applicable)
* Review handler responses match the Discord embed/button format documented in `docs/technical-documentation.md`

## Checklist 
- [x] Code builds successfully
- [ ] Tests added or updated
- [ ] Documentation updated/added
- [x] No breaking changes

## Screenshots (if applicable)

N/A — Discord bot interactions (no web UI)

### CODE REVIEW

File: Task2_mhp_discord-bot/backend/src/handlers/location_handler.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
+# Handles /work-location slash command and location button interactions
+import logging
+from datetime import date
+from typing import Any, Dict
+
+from src.handlers.auth import AuthenticatedUser
+from src.models import WorkLocationType
+from src.services.location_service import location_service, LOCATION_DISPLAY
```
Comment (ghalib-craftsmen) [inline Task2_mhp_discord-bot/backend/src/handlers/location_handler.py:8]: LOCATION_DISPLAY is imported but never referenced anywhere in the file. The handler correctly uses location_service.get_location_display_info() instead. Remove the unused name.

File: Task2_mhp_discord-bot/backend/src/services/location_service.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                f"⚠️ You have already used {wfh_count}/{cap} WFH days this month "
+                f"({target_date.strftime('%B %Y')}). Monthly soft limit reached."
+            )
+
+        return False, None
+
+    def _get_user_wfh_days_in_month(
+        self, discord_id: str, target_date: date
+    ) -> list[WorkLocation]:
+        import calendar
```
Comment (ghalib-craftsmen) [inline Task2_mhp_discord-bot/backend/src/services/location_service.py:113]: calendar is a stdlib module with no import cost. Can we move it to the top-level imports? Inline imports are only justified for optional heavy dependencies, I think this is not that case.

File: Task2_mhp_discord-bot/backend/src/services/location_service.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        return results
+
+    # ── Display helpers ──────────────────────────────────────────────────────
+
+    def get_location_display_info(self, loc_type: WorkLocationType) -> dict:
+        return LOCATION_DISPLAY.get(
+            loc_type,
+            {"label": loc_type.value.upper(), "emoji": "📍"},
+        )
+location_service = LocationService()
```
Comment (ghalib-craftsmen) [inline Task2_mhp_discord-bot/backend/src/services/location_service.py:135]: The function is glued directly to the last line of the class. Add two blank lines at least after the function end to separate it and make it clearer that it’s a module-level object.

**Review by ghalib-craftsmen — CHANGES_REQUESTED** [review]

File: Task2_mhp_discord-bot/backend/src/handlers/location_handler.py (hunk 1 above)
Comment (NiloyGabrielGomes) [reply Task2_mhp_discord-bot/backend/src/handlers/location_handler.py:8]: Removed in later iterations.

File: Task2_mhp_discord-bot/backend/src/services/location_service.py (hunk 2 above)
Comment (NiloyGabrielGomes) [reply Task2_mhp_discord-bot/backend/src/services/location_service.py:113]: Fixed in later iteration. PR #62

**Review by ghalib-craftsmen — APPROVED** [review]
